### ISSUE TITLE
TFKeras-Integration along with ModelCheckpoint that only saves best weights

### ISSUE THREAD

**Issue by MarkusMiller** [optuna/optuna#2498]
Hello,

I recently switched from Talos to Optuna and so far it's running great. So I set up my code with Talos to first find good hyperparameters in a base training run (over one to five epochs, picking the combination with the lowest loss at the end) and then did a fine-tuning training with those parameters over several epochs. I sort of incorporated Optuna now into this approach, however, I'm having a small issue, because I used a Keras ModelCheckpoint Callback that is set to only store the best weights, which is some sort of inherent early stopping. So, for example, if the Trial is running for 5 Epochs and the best value to achieve was on Epoch 3, Optuna reports the last known loss, which is the one from Epoch 5, as the best loss, ignoring Epoch 3, on which the loss was better than the last. 

Is there anything I could do, so Optuna would return said 3rd epoch result instead of the final one?

Thank you very much and kind regards,

Markus

**Comment by crcrpar** [comment]
Hi @MarkusMiller 

I'm glad to hear that you are using Optuna.
I think you can use the ModelCheckpoint Callback instance's best value as a return value of `objective` instead of the latest metrics value. 

I created a simple example with the above setting: https://colab.research.google.com/drive/18QdQHgYn2-PN-f1V3gOr1x-fPjmBIHAc?usp=sharing.

I hope this helps you a bit.

**Comment by MarkusMiller** [comment]
Hello, 

thank you for your very quick reply. I guess, that'll solve my problem. Thank you very much.

Kind regards

Markus

**Comment by crcrpar** [comment]
Glad to hear that. Let me close this issue.

Should you have further questions or issues, feel free to reopen / create :)
